db/mongo.go

```diff
@@ -194,19 +200,73 @@ func (db *database) buildRestoreArgs(dumpPath string, includeStatusHistory bool)
 
 // RestoreFromDump uses mongorestore to load the dump from a backup.
 func (db *database) RestoreFromDump(dumpDir, logFile string, includeStatusHistory bool) error {
+	binary, isSnap, err := db.getRestoreBinary()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Snap mongorestore can only access certain directories, so move the dump
+	// from /tmp to under $HOME/snap before running restore, and delete after.
+	if isSnap {
+		dumpDir, err = db.moveToHomeSnap(dumpDir)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer func() {
+			err := os.RemoveAll(dumpDir)
+			if err != nil {
+				logger.Warningf("error removing snap dump dir: %v", err)
+			}
+		}()
+	}
+
 	command := exec.Command(
-		restoreBinary,
+		binary,
 		db.buildRestoreArgs(dumpDir, includeStatusHistory)...,
 	)
-	logger.Debugf("running restore command: %s %s", command.Path, strings.Join(command.Args, " "))
-	dest, err := os.Create(logFile)
+	logger.Debugf("running restore command: %s", strings.Join(command.Args, " "))
+
+	// Use CombinedOutput and then write the bytes ourselves instead of
+	// passing a file for command.Stdout/Stderr -- this avoids a permissions
+	// issue with the Snap mongorestore writing to the file.
+	output, err := command.CombinedOutput()
+	if err != nil {
+		logger.Debugf("%s output:\n%s", binary, output)
+		return errors.Annotatef(err, "running %s", binary)
+	}
+	err = ioutil.WriteFile(logFile, output, 0664)
+	if err != nil {
+		logger.Debugf("%s output:\n%s", binary, output)
+		return errors.Annotatef(err, "writing output to %s", logFile)
+	}
+	return nil
+}
+
+func (db *database) getRestoreBinary() (binary string, isSnap bool, err error) {
+	if _, err := exec.LookPath(snapRestoreBinary); err == nil {
+		return snapRestoreBinary, true, nil
+	}
+	if _, err := exec.LookPath(restoreBinary); err == nil {
+		return restoreBinary, false, nil
+	}
+	return "", false, errors.Errorf("couldn't find %s or %s in PATH (%s)",
+		snapRestoreBinary, restoreBinary, os.Getenv("PATH"))
+}
+
+func (db *database) moveToHomeSnap(dumpDir string) (string, error) {
+	snapDumpDir := filepath.Join(os.ExpandEnv(homeSnapDir), dumpDir)
```

Comment:
we should check $HOME is set, if not we should error.

Reply:
if we don't, it would move to /snap/juju-db/common which is bad

Reply:
Good call -- I switched to using `os.UserHomeDir()` instead (and checking its error, which is non-nil if HOME is not set). That function is also more portable, for what that's worth.